File: backend/app/security/kdf.py
"""Client-side key schedule (reference implementation).

    master_key = PBKDF2-HMAC-SHA256(password, salt, 600_000)
    auth_key   = HKDF-SHA256(master_key, info="mindpattern/auth/v1")   # sent to server
    data_key   = HKDF-SHA256(master_key, info="mindpattern/data/v1")   # NEVER sent,
                                                       # except into a processing session

The server stores scrypt(auth_key) and can therefore verify logins without
ever holding a key that decrypts entries. The mobile client derives the same
three values locally (shared/vectors.json pins the exact bytes on both
platforms).
"""

from __future__ import annotations

import hashlib
import hmac

KDF_ITERATIONS = 600_000
AUTH_INFO = b"mindpattern/auth/v1"
DATA_INFO = b"mindpattern/data/v1"
MIN_SALT_SIZE = 8


def derive_master_key(
    password: str | bytes, salt: bytes, iterations: int = KDF_ITERATIONS
) -> bytes:
    """Derive the 256-bit master key from the user's password."""
    if isinstance(password, str):
        password = password.encode("utf-8")
    if len(salt) < MIN_SALT_SIZE:
        raise ValueError(f"salt must be at least {MIN_SALT_SIZE} bytes")
    if iterations < 1:
        raise ValueError("iterations must be positive")
    return hashlib.pbkdf2_hmac("sha256", password, salt, iterations)


def hkdf_sha256(ikm: bytes, salt: bytes | None, info: bytes, length: int = 32) -> bytes:
    """HKDF-SHA256 (RFC 5869) extract-and-expand."""
    if length < 1 or length > 255 * 32:
        raise ValueError("invalid HKDF output length")
    salt = salt if salt else b"\x00" * hashlib.sha256().digest_size
    prk = hmac.new(salt, ikm, hashlib.sha256).digest()
    okm = b""
    block = b""
    counter = 1
    while len(okm) < length:
        block = hmac.new(prk, block + info + bytes([counter]), hashlib.sha256).digest()
        okm += block
        counter += 1
    return okm[:length]


def derive_auth_key(master_key: bytes) -> bytes:
    """Key whose scrypt hash the server stores for login verification."""
    return hkdf_sha256(master_key, None, AUTH_INFO)


def derive_data_key(master_key: bytes) -> bytes:
    """Key that encrypts entries, insights and questions."""
    return hkdf_sha256(master_key, None, DATA_INFO)

File: mobile/src/crypto/kdf.ts
/**
 * Client-side key schedule — must match backend/app/security/kdf.py exactly.
 *
 *   master_key = PBKDF2-HMAC-SHA256(password, salt, 600_000)
 *   auth_key   = HKDF-SHA256(master_key, info="mindpattern/auth/v1")  -> sent to server
 *   data_key   = HKDF-SHA256(master_key, info="mindpattern/data/v1")  -> never leaves
 *                                                            the device, except into a
 *                                                            TLS-delivered, single-use
 *                                                            processing session.
 *
 * Byte-for-byte compatibility is pinned by shared/vectors.json; run
 * `npm test` (the real modules against the vectors) — it replaces the old
 * webcrypto-only reference check.
 */
import { engine } from "./engine";

export const KDF_ITERATIONS = 600_000;
const MIN_SALT_SIZE = 8;

// HKDF info strings are rebuilt at call time, not captured at module scope:
// every derivation then exercises (and mutation-tests) the exact label
// bytes, and no warm test worker can cache a stale value.
function authInfo(): Buffer {
  // "utf8" is Buffer's default; an empty encoding string decodes identically.
// Stryker disable StringLiteral
return Buffer.from("mindpattern/auth/v1", "utf8");
// Stryker restore StringLiteral
}

function dataInfo(): Buffer {
  // Stryker disable StringLiteral
return Buffer.from("mindpattern/data/v1", "utf8");
  // Stryker restore StringLiteral
}

export function deriveMasterKey(password: string, salt: Buffer, iterations = KDF_ITERATIONS): Buffer {
  if (salt.length < MIN_SALT_SIZE) {
    throw new Error(`salt must be at least ${MIN_SALT_SIZE} bytes`);
  }
  if (iterations < 1) {
    throw new Error("iterations must be positive");
  }
  return engine.pbkdf2Sync(password, salt, iterations, 32, "sha256");
}

/**
 * The async derivation — PREFER this in UI paths. The sync form freezes
 * the JS thread for ~100-400ms at 600k iterations; the engine's async
 * pbkdf2 (quick-crypto's JSI worker on device, node:crypto's thread pool
 * in tests) keeps the UI responsive. Same bytes, same vectors.
 */
export async function deriveMasterKeyAsync(
  password: string,
  salt: Buffer,
  iterations = KDF_ITERATIONS,
): Promise<Buffer> {
  if (salt.length < MIN_SALT_SIZE) {
    throw new Error(`salt must be at least ${MIN_SALT_SIZE} bytes`);
  }
  if (iterations < 1) {
    throw new Error("iterations must be positive");
  }
  return new Promise((resolve, reject) => {
    engine.pbkdf2(password, salt, iterations, 32, "sha256", (err, derivedKey) => {
      if (err) {
        reject(err);
        return;
      }
      if (!derivedKey) {
        reject(new Error("pbkdf2 produced no key"));
        return;
      }
      resolve(Buffer.from(derivedKey));
    });
  });
}

export function deriveAuthKey(masterKey: Buffer): Buffer {
  // Empty salt -> RFC 5869 default of HashLen zeros, matching the backend.
  return hkdfSha256(masterKey, authInfo());
}

export function deriveDataKey(masterKey: Buffer): Buffer {
  return hkdfSha256(masterKey, dataInfo());
}

/**
 * hkdfSync returns an ArrayBuffer (Node API parity in quick-crypto), not a
 * Buffer — wrap it so Buffer methods (toString("base64"), fill, subarray)
 * actually exist on the result.
 */
function hkdfSha256(ikm: Buffer, info: Buffer): Buffer {
  const derived = engine.hkdfSync("sha256", ikm, Buffer.alloc(32), info, 32);
  return Buffer.from(derived as ArrayBuffer);
}

/** Overwrite key material buffers as soon as they are no longer needed. */
export function zeroize(...buffers: (Buffer | undefined | null)[]): void {
  for (const buf of buffers) {
    if (buf) buf.fill(0);
  }
}
